server/src/service.rs

```diff
@@ -1,30 +1,76 @@
 // org.freedesktop.Secret.Service
 
-use std::collections::HashMap;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
 
-use oo7::dbus::{
-    api::{Properties, SecretInner},
-    Algorithm, ServiceError,
+use oo7::{
+    dbus::{
+        api::{Properties, SecretInner},
+        Algorithm, ServiceError,
+    },
+    Key,
 };
+use tokio::sync::RwLock;
 use zbus::{
     proxy::ProxyDefault,
     zvariant::{ObjectPath, OwnedObjectPath, OwnedValue, Value},
 };
 
+use crate::{service_manager::ServiceManager, session::Session};
+
 pub type Result<T> = std::result::Result<T, ServiceError>;
 
 #[derive(Debug, Default)]
-pub struct Service {}
+pub struct Service {
+    manager: Arc<Mutex<ServiceManager>>,
+    session_index: RwLock<i32>,
+}
 
 #[zbus::interface(name = "org.freedesktop.Secret.Service")]
 impl Service {
     #[zbus(out_args("output", "result"))]
     pub async fn open_session(
         &self,
-        _algorithm: Algorithm,
-        _input: Value<'_>,
+        algorithm: Algorithm,
+        input: Value<'_>,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
     ) -> Result<(OwnedValue, OwnedObjectPath)> {
-        todo!()
+        let (public_key, aes_key) = match algorithm {
+            Algorithm::Plain => (None, None),
+            Algorithm::Encrypted => {
+                let client_public_key = Key::from(input);
+                let private_key = Key::generate_private_key();
+                (
+                    Some(Key::generate_public_key(&private_key)),
+                    Some(Key::generate_aes_key(&private_key, &client_public_key)),
+                )
+            }
+        };
+
+        let mut sessions = *self.session_index.read().await;
+        sessions += 1;
+        *self.session_index.write().await = sessions;
+
+        let session = Session::new(aes_key.map(Arc::new), Arc::clone(&self.manager), sessions);
+        let path = session.path().clone();
```

Comment:
Why clone here?

Reply:
Yeah I agree. It doesn't make sense to clone here (but below). But when removed, I'm getting a lengthy error,

```
error[E0277]: the trait bound `zbus::zvariant::ObjectPath<'_>: From<&&OwnedObjectPath>` is not satisfied
  --> server/src/service.rs:70:23
   |
70 |         object_server.at(&path, session).await?;
   |                       ^^ the trait `From<&&OwnedObjectPath>` is not implemented for `zbus::zvariant::ObjectPath<'_>`, which is required by `zbus::zvariant::ObjectPath<'_>: TryFrom<&&OwnedObjectPath>`
   |
   = help: the following other types implement trait `From<T>`:
             `zbus::zvariant::ObjectPath<'o>` implements `From<&zbus::zvariant::ObjectPath<'o>>`
             `zbus::zvariant::ObjectPath<'static>` implements `From<OwnedObjectPath>`
             `zbus::zvariant::ObjectPath<'unowned>` implements `From<&'owned OwnedObjectPath>`
   = note: required for `&&OwnedObjectPath` to implement `Into<zbus::zvariant::ObjectPath<'_>>`
   = note: required for `zbus::zvariant::ObjectPath<'_>` to implement `TryFrom<&&OwnedObjectPath>`

error[E0277]: the trait bound `zbus::zvariant::ObjectPath<'_>: From<&&OwnedObjectPath>` is not satisfied
  --> server/src/service.rs:70:9
   |
70 |         object_server.at(&path, session).await?;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<&&OwnedObjectPath>` is not implemented for `zbus::zvariant::ObjectPath<'_>`, which is required by `zbus::zvariant::ObjectPath<'_>: TryFrom<&&OwnedObjectPath>`
   |
   = help: the following other types implement trait `From<T>`:
             `zbus::zvariant::ObjectPath<'o>` implements `From<&zbus::zvariant::ObjectPath<'o>>`
             `zbus::zvariant::ObjectPath<'static>` implements `From<OwnedObjectPath>`
             `zbus::zvariant::ObjectPath<'unowned>` implements `From<&'owned OwnedObjectPath>`
   = note: required for `&&OwnedObjectPath` to implement `Into<zbus::zvariant::ObjectPath<'_>>`
   = note: required for `zbus::zvariant::ObjectPath<'_>` to implement `TryFrom<&&OwnedObjectPath>`

error[E0277]: the trait bound `zbus::zvariant::ObjectPath<'_>: From<&&OwnedObjectPath>` is not satisfied
  --> server/src/service.rs:70:42
   |
70 |         object_server.at(&path, session).await?;
   |                                          ^^^^^ the trait `From<&&OwnedObjectPath>` is not implemented for `zbus::zvariant::ObjectPath<'_>`, which is required by `zbus::zvariant::ObjectPath<'_>: TryFrom<&&OwnedObjectPath>`
   |
   = help: the following other types implement trait `From<T>`:
             `zbus::zvariant::ObjectPath<'o>` implements `From<&zbus::zvariant::ObjectPath<'o>>`
             `zbus::zvariant::ObjectPath<'static>` implements `From<OwnedObjectPath>`
             `zbus::zvariant::ObjectPath<'unowned>` implements `From<&'owned OwnedObjectPath>`
   = note: required for `&&OwnedObjectPath` to implement `Into<zbus::zvariant::ObjectPath<'_>>`
   = note: required for `zbus::zvariant::ObjectPath<'_>` to implement `TryFrom<&&OwnedObjectPath>`

For more information about this error, try `rustc --explain E0277`.

```

Reply:
ah! figured out the error.

Reply:
so, `object_server.at(path.clone(), session).await?;` call moves `session`, that's why I have cloned the path here.